[plsql] Parsing exception "Less than or equal to/Greater than or equal to" operators in DML statements

**Affects PMD Version:** 

This error was testing in the following versions:

- 6.41.0
- 6.40.0
- 6.39.0

**Description:**

Trying to analyze pl/sql code that has "Less than or equal to/Greater than or equal to" operators (<=; >=), but these operators have a blank space between two characters, for instance  > = or < =, PMD raises an exception error indicating "Error while parsing". 

_Note:_ This is totally correct in DML statements for Oracle databases engines.

**Exception Stacktrace:**

```
net.sourceforge.pmd.PMDException: Error while parsing C:\testfolder\example_procedure.prc
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCodeWithoutCache(SourceCodeProcessor.java:124)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:100)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:62)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:85)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:29)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: net.sourceforge.pmd.lang.plsql.ast.ParseException: Encountered " "=" "= "" at line 10, column 23.
Was expecting one of:
    "LOG" ...
    "(" ...
    "REF" ...
    "LAST" ...
    "*" ...
    "+" ...
    "OVER" ...
    "UNBOUNDED" ...
    "PRECEDING" ...
    ...
    "TRIM" ...
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.generateParseException(PLSQLParser.java)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.jj_consume_token(PLSQLParser.java)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.UnaryExpression(PLSQLParser.java:41132)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.MultiplicativeExpression(PLSQLParser.java:40679)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.AdditiveExpression(PLSQLParser.java:40565)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.SqlExpression(PLSQLParser.java:13560)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.ComparisonCondition(PLSQLParser.java:12848)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Condition2(PLSQLParser.java:12165)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.CompoundCondition(PLSQLParser.java:12703)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Condition(PLSQLParser.java:11754)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.CompoundCondition(PLSQLParser.java:12675)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Condition(PLSQLParser.java:11754)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.CompoundCondition(PLSQLParser.java:12697)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Condition(PLSQLParser.java:11754)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.CompoundCondition(PLSQLParser.java:12675)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Condition(PLSQLParser.java:11754)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.CompoundCondition(PLSQLParser.java:12697)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Condition(PLSQLParser.java:11754)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.CompoundCondition(PLSQLParser.java:12697)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Condition(PLSQLParser.java:11754)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.CompoundCondition(PLSQLParser.java:12675)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Condition(PLSQLParser.java:11754)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.CompoundCondition(PLSQLParser.java:12724)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Condition(PLSQLParser.java:11754)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.CompoundCondition(PLSQLParser.java:12724)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Condition(PLSQLParser.java:11754)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.CompoundCondition(PLSQLParser.java:12724)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Condition(PLSQLParser.java:11754)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.WhereClause(PLSQLParser.java:11712)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.RestOfStatement(PLSQLParser.java:10606)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.SelectStatement(PLSQLParser.java:28448)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.CursorUnit(PLSQLParser.java:38877)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.CursorSpecification(PLSQLParser.java:38919)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.DeclarativeUnit(PLSQLParser.java:2728)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.DeclarativeSection(PLSQLParser.java:3154)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.ProgramUnit(PLSQLParser.java:6446)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Global(PLSQLParser.java:879)
	at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Input(PLSQLParser.java:153)
	at net.sourceforge.pmd.lang.plsql.PLSQLParser.parse(PLSQLParser.java:61)
	at net.sourceforge.pmd.lang.AbstractParser.doParse(AbstractParser.java:44)
	at net.sourceforge.pmd.SourceCodeProcessor.parse(SourceCodeProcessor.java:136)
	at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:200)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCodeWithoutCache(SourceCodeProcessor.java:118)
	... 10 more
```

**Code Sample demonstrating the issue:**

```
CREATE OR REPLACE PROCEDURE EXAMPLE_PROCEDURE IS
   --
   l_total_objects NUMBER(10);
   --
BEGIN
   --
   SELECT COUNT(1)
     INTO l_total_objects
     FROM USER_OBJECTS
    WHERE namespace < = 1
       OR namespace > = 0;
   --
   DBMS_OUTPUT.Put_Line('Total number of objects: ' || l_total_objects);
   --
END EXAMPLE_PROCEDURE;
```

_Note:_ By removing all the blank spaces in the operators in the SELECT statement, you will have no parsing problems.

**Steps to reproduce:**

1.  Put the code above ("Code Sample demonstrating the issue" section) in example_procedure.prc
2.  Execute PMD analyzer from CLI.
3. See the error showed in the "Exception Stacktrace" section.

**Running PMD through:** *[CLI]*

--- a/pmd-plsql/etc/grammar/PldocAST.jjt
+++ b/pmd-plsql/etc/grammar/PldocAST.jjt
@@ -257,6 +257,34 @@ TOKEN_MGR_DECLS : {
   List<ParsingExclusion> exclusions = new ArrayList<ParsingExclusion>();
 }
 
+// add names to SINGLE CHARACTER tokens
+// multiple character operators are sequences of tokens.
+TOKEN: {
+  < LPAREN: "(" >
+| < RPAREN: ")" >
+| < LBRACE: "{" >
+| < RBRACE: "}" >
+| < LBRACKET: "[" >
+| < RBRACKET: "]" >
+| < SEMICOLON: ";" >
+| < COMMA: "," >
+| < DOT: "." >
+| < AROBASE: "@" >
+| < ASSIGN: "=" >
+| < LT: "<" >
+| < BANG: "!" >
+| < TILDE: "~" >
+| < HOOK: "?" >
+| < COLON: ":" >
+| < PLUSSIGN: "+" >
+| < MINUSSIGN: "-" >
+| < STAR: "*" >
+| < SLASH: "/" >
+| < BIT_AND: "&" >
+| < BIT_OR: "|" >
+| < XOR: "^" >
+| < REM: "%" >
+}
 
 /**
  * 2006-05-22 - Matthias Hendler - Added parsing of triggers and global functions/procedures
@@ -1603,18 +1631,26 @@ ASTInCondition InCondition() :
     { return jjtThis; }
 }
 
-ASTComparisonCondition ComparisonCondition() :
+private String ComparisonOp() #void :
 {}
 {
-    (
-    // SimpleComparisonCondition
-      LOOKAHEAD(3) (SqlExpression() ( "=" | "!=" | "^=" | "<>" | ">" | "<" | ">=" | "<=" ) { jjtThis.setOperator(token.getImage()); } SqlExpression() )
-    | LOOKAHEAD(3) ( "(" SqlExpression() ("," SqlExpression())* ")" ( "=" | "!=" | "^=" | "<>" ) { jjtThis.setOperator(token.getImage()); } "(" ( LOOKAHEAD(3) ExpressionList() | LOOKAHEAD(3) Subquery() ) ")" )
+      "=" {return "=";}
+    | "!" "=" {return "!=";}
+    | "^" "=" {return "^=";}
+    | "<" ( "=" {return "<=";} | ">" {return "<>";} )? {return "<";}
+    | ">" ( "=" {return ">=";} )? {return ">";}
+}
 
-    // GroupComparisonCondition
-    | LOOKAHEAD(3) (SqlExpression() ( "=" | "!=" | "^=" | "<>" | ">" | "<" | ">=" | "<=" ) { jjtThis.setOperator(token.getImage()); } ( <ANY> | <SOME> | <ALL> ) "(" ( LOOKAHEAD(3) ExpressionListSingle() | LOOKAHEAD(3) Subquery() ) ")" )
-    | LOOKAHEAD(3) ( "(" SqlExpression() ("," SqlExpression())* ")" ( "=" | "!=" | "^=" | "<>" ) { jjtThis.setOperator(token.getImage()); } ( <ANY> | <SOME> | <ALL> ) "(" ( LOOKAHEAD(3) ExpressionListMultiple() ("," ExpressionListMultiple())* | LOOKAHEAD(3) Subquery() ) ")" )
-    )
+ASTComparisonCondition ComparisonCondition() :
+{String op;}
+{
+    (LOOKAHEAD(3) SqlExpression() | ExpressionListMultiple() )
+    op=ComparisonOp() { jjtThis.setOperator(op); }
+    ( <ANY> | <SOME> | <ALL> )?
+    (LOOKAHEAD(3) "(" Subquery() ")"
+    |LOOKAHEAD(3) "(" ExpressionListSingle() ")"
+    |LOOKAHEAD(3) "(" ExpressionListMultiple() ("," ExpressionListMultiple())* ")"
+    |LOOKAHEAD(3) SqlExpression())
     { return jjtThis; }
 }
 
@@ -3179,17 +3215,17 @@ ASTEqualityExpression EqualityExpression() #EqualityExpression(>1) :
 {
   //RelationalExpression() ( ( "=" | "!=" | "<>" | <IS>) RelationalExpression() )*
   (
+
+	  simpleNode=RelationalExpression()
+	  {sb.append(simpleNode.getImage());}
 	  (
-	    simpleNode = RelationalExpression()
-	  ) {sb.append(simpleNode.getImage());}
-	  (
-	   ( ("=" ) {sb.append(" = ");}
-	   | ("!" "=" ) {sb.append(" != ");}
+	   ( "=" {sb.append(" = ");}
+	   | "!" "=" {sb.append(" != ");}
 	   // Now unde RelationalExpression | ("<" ">" ) {sb.append(" <> ");}
-	   | (<IS>) {sb.append(" IS ");})
-	   (
-	   simpleNode = RelationalExpression()
-	   ) {sb.append(simpleNode.getImage());}
+	   | <IS> {sb.append(" IS ");}
+	   )
+	   simpleNode=RelationalExpression()
+	   {sb.append(simpleNode.getImage());}
 	  )*
   )
  {
@@ -3209,17 +3245,16 @@ ASTRelationalExpression RelationalExpression() #RelationalExpression(>1) :
   (
     (
 
-      (
-          ("<>" { sb.append(" <> "); } )
-        | ("<=" { sb.append(" <= "); } )
-        | (">=" { sb.append(" >= "); } )
-        | ("!=" { sb.append(" != "); } )
-        | ("~=" { sb.append(" ~= "); } )
-        | ("^=" { sb.append(" ^= "); } )
-        | ("<" { sb.append(" < "); } )
-        | (">" { sb.append(" > "); } )
+      (  LOOKAHEAD(2) "<" ">" { sb.append(" <> "); }
+        |LOOKAHEAD(2) "<" "=" { sb.append(" <= "); }
+        |LOOKAHEAD(2) ">" "=" { sb.append(" >= "); }
+        // then lookahead(1) is enough
+        | "<" { sb.append(" < "); }
+        | ">" { sb.append(" > "); }
+        | "!" "=" { sb.append(" != "); }
+        | "~" "=" { sb.append(" ~= "); }
+        | "^" "=" { sb.append(" ^= "); }
       )
-
       { sb.append(" "); }
       | [(<NOT>) { sb.append(" NOT "); }]
       (

--- a/pmd-plsql/src/main/java/net/sourceforge/pmd/lang/plsql/PLSQLParser.java
+++ b/pmd-plsql/src/main/java/net/sourceforge/pmd/lang/plsql/PLSQLParser.java
@@ -69,3 +69,4 @@ public class PLSQLParser extends AbstractParser {
         return new HashMap<>(); // FIXME
     }
 }
+

--- a/pmd-plsql/src/test/java/net/sourceforge/pmd/lang/plsql/ast/PLSQLParserTest.java
+++ b/pmd-plsql/src/test/java/net/sourceforge/pmd/lang/plsql/ast/PLSQLParserTest.java
@@ -75,6 +75,12 @@ public class PLSQLParserTest extends AbstractPLSQLParserTst {
     }
 
     @Test
+    public void testRelationalOperators() {
+        // https://github.com/pmd/pmd/issues/3746
+        plsql.parseResource("RelationalOperators.pls");
+    }
+
+    @Test
     public void testExecuteImmediateIssue3106() {
         plsql.parseResource("ExecuteImmediateIssue3106.pls");
     }

--- a/pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/RelationalOperators.pls
+++ b/pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/RelationalOperators.pls
@@ -1,0 +1,18 @@
+CREATE OR REPLACE PROCEDURE EXAMPLE_PROCEDURE IS
+   --
+   l_total_objects NUMBER(10);
+   --
+BEGIN
+   --
+   SELECT COUNT(1)
+     INTO l_total_objects
+     FROM USER_OBJECTS
+    WHERE namespace < = 1
+       OR namespace > = 0
+       OR namespace >
+       = 0
+       ;
+   --
+   DBMS_OUTPUT.Put_Line('Total number of objects: ' || l_total_objects);
+   --
+END EXAMPLE_PROCEDURE;

